java/arcs/core/host/EntityHandleManager.kt

```diff
@@ -0,0 +1,452 @@
+     *
+     * @property handleHolder contains handle and entitySpec declarations
+     * @property handleName name for the handle, must be present in [HandleHolder.entitySpecs]
+     * @property storageKey a [StorageKey]
+     * @property schema the [Schema] for this [StorageKey]
+     * @property handleMode whether a handle is Read,Write, or ReadWrite (default)
+     * @property sender block used to execute callback lambdas
+     * @property idGenerator used to generate unique IDs for newly stored entities.
+     */
+    suspend fun createSetHandle(
+        handleHolder: HandleHolder,
+        handleName: String,
+        storageKey: StorageKey,
+        schema: Schema,
+        handleMode: HandleMode = HandleMode.ReadWrite,
+        idGenerator: Id.Generator = Id.Generator.newSession(),
+        sender: Sender = ::defaultSender
+    ) = createSdkHandle(
+        handleHolder,
+        handleName,
+        handleManager.setHandle(storageKey, schema, canRead = handleMode != HandleMode.Write),
+        handleMode,
+        idGenerator,
+        sender
+    )
+
+    /**
+     * Same-thread non-blocking dispatch. Note that this may lead to concurrency problems on
+     * some platforms. On platforms with real preemptive threads, use an implementation that
+     * provides concurrency aware dispatch.
+     */
+    private fun defaultSender(block: suspend () -> Unit) {
+        GlobalScope.launch {
+            block()
+        }
+    }
+
+    /**
+     * Create a [Handle] given a [StorageHandle] and a [EntitySpec] definition.
+     *
+     * @property entitySpec used to deserialize [RawEntity] types into [Entity]
+     * @property handleName readable name for the handle, usually from a recipe
+     * @property storageHandle a [StorageHandle] usually obtained thru [HandleManager]
+     * @property handleMode whether a handle is Read,Write, or ReadWrite (default)
+     * @property sender block used to execute callback lambdas
+     * @property idGenerator used to generate unique IDs for newly stored entities.
+     */
+    private fun <T : Entity> createSdkHandle(
+        entitySpec: EntitySpec<T>,
+        handleName: String,
+        storageHandle: StorageHandle<*, *, *>,
+        handleMode: HandleMode,
+        idGenerator: Id.Generator,
+        sender: Sender
+    ): Handle {
+        return when (storageHandle) {
+            is SingletonHandle<*> -> createSingletonHandle(
+                entitySpec,
+                handleName,
+                storageHandle as SingletonHandle<RawEntity>,
+                handleMode,
+                idGenerator,
+                sender
+            )
+            is SetHandle<*> -> createSetHandle(
+                entitySpec,
+                handleName,
+                storageHandle as SetHandle<RawEntity>,
+                handleMode,
+                idGenerator,
+                sender
+            )
+            else -> throw Exception("Unknown storage handle type ${storageHandle::class}")
```

Comment:
Please don't throw base Exceptions in the future.

Reply:
@jasonwyatt Got any advice on which exception class to use when? Kotlin has a different set to the ones I'm used to.